### PR TITLE
Give each redshift client their own boto session

### LINKED ISSUE #2756: dbt compile fails against redshift when using multi-threading
### Describe the bug
When I run `dbt compile` against our redshift data warehouse the command fails with the error 
```
KeyError: 'endpoint_resolver'
```
The error only occurs when threads > 1 and `method: iam` is used.

From what I can gather this is happening because the boto session object is not thread safe and it is being accessed from multiple threads without being protected, and this happens during the call to get_tmp_iam_cluster_credentials.

### Steps To Reproduce
* Create a dbt project containing a significant number of models.
* Configure the target in the profiles.yml file to point to a redshift dwh, with threads > 1 and method iam:
```
      type: redshift
      method: iam
      threads: 8
      host: xxxx.redshift.amazonaws.com
      cluster_id: xxx
      port: 5439
      user: xxx
      dbname: xxx
      schema: xxx
```
* Run `dbt compile`

### Expected behavior
That `dbt compile` succeeds.

### Screenshots and log output

```
2020-09-14 11:15:23.743840 (MainThread): Traceback (most recent call last):
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/postgres/connections.py", line 46, in exception_handler
    yield
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/sql/connections.py", line 76, in add_query
    cursor = connection.handle.cursor()
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/contracts/connection.py", line 69, in handle
    self._handle.resolve(self)
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/contracts/connection.py", line 90, in resolve
    return self.opener(connection)
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/postgres/connections.py", line 77, in open
    credentials = cls.get_credentials(connection.credentials)
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/redshift/connections.py", line 152, in get_credentials
    return cls.get_tmp_iam_cluster_credentials(credentials)
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/redshift/connections.py", line 128, in get_tmp_iam_cluster_credentials
    credentials.db_groups,
  File "/Users/xxx/venv/lib/python3.7/site-packages/dbt/adapters/redshift/connections.py", line 93, in fetch_cluster_credentials
    boto_client = boto3.client('redshift')
  File "/Users/xxx/venv/lib/python3.7/site-packages/boto3/__init__.py", line 91, in client
    return _get_default_session().client(*args, **kwargs)
  File "/Users/xxx/venv/lib/python3.7/site-packages/boto3/session.py", line 263, in client
    aws_session_token=aws_session_token, config=config)
  File "/Users/xxx/venv/lib/python3.7/site-packages/botocore/session.py", line 828, in create_client
    endpoint_resolver = self._get_internal_component('endpoint_resolver')
  File "/Users/xxx/venv/lib/python3.7/site-packages/botocore/session.py", line 695, in _get_internal_component
    return self._internal_components.get_component(name)
  File "/Users/xxx/venv/lib/python3.7/site-packages/botocore/session.py", line 907, in get_component
    del self._deferred[name]
KeyError: 'endpoint_resolver'
```

Sometimes the error returned is 
```
KeyError: 'credential_provider'
```
but the stack trace is identical.

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [x] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.17.2
   latest version: 0.18.0

Your version of dbt is out of date! You can find instructions for upgrading here:
https://docs.getdbt.com/docs/installation

Plugins:
  - bigquery: 0.17.2
  - snowflake: 0.17.2
  - redshift: 0.17.2
  - postgres: 0.17.2
```

**The operating system you're using:**
macOS Catalina

**The output of `python --version`:**
Python 3.7.3

### Additional context
The error surfaced after I bumped dbt from version 0.14.2 to 0.17.2

### PR DESCRIPTION
**Pull request by jweibel22**
resolves #2756

### Description
Since the boto session is not thread-safe, using the default session in a multi-threaded scenario will result in concurrency errors.
This change ensures that a fresh session is instantiated before creating the client. This is similar to what is happening in the else branch. Thus, this change should not cause any unexpected issues (e.g. exceeding service quotas)

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by jtcohen6** [comment]
@jweibel22 Nice! I'd love to get this in for v0.18.1. Does this change work for you locally when you run with multiple threads?

Also, could you:
- pull changes from `dev/0.18.1` (which should include fixes to the failing Redshift integration tests)
- add a changelog entry under v0.18.1

**Comment by jweibel22** [comment]
> @jweibel22 Nice! I'd love to get this in for v0.18.1. Does this change work for you locally when you run with multiple threads?

Yes, I have verified that is does resolve the issue when I run `dbt compile` locally.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 ## dbt 0.18.1 (Release TBD)
 
+### Fixes
+- `dbt compile` and `dbt run` failed with `KeyError: 'endpoint_resolver'` when threads > 1 and `method: iam` had been specified in the profiles.yaml ([#2766](https://github.com/fishtown-analytics/dbt/pull/2766))
```
Comment (beckjake) [inline CHANGELOG.md:4]:
```suggestion
- `dbt compile` and `dbt run` failed with `KeyError: 'endpoint_resolver'` when threads > 1 and `method: iam` had been specified in the profiles.yaml ([#2756](https://github.com/fishtown-analytics/dbt/issues/2756), [#2766](https://github.com/fishtown-analytics/dbt/pull/2766))
```

We usually also include a link to the issue in the changelog

**Review by beckjake — CHANGES_REQUESTED** [review]
This looks very nice!

Could you add a test to the redshift unit tests in `tests/unit/test_redshift_adapter.py` that makes sure we call `boto3.Session().client()`  instead of `boto3.client()` directly in both the iam_profile  cases (none and not none)? I would hate to regress on this.

**Review by beckjake — APPROVED** [review]
Thanks! This looks great to me

**Review by jtcohen6 — APPROVED** [review]
Thanks for this @jweibel22! I'll try to resolve the changelog conflict and get this in.

**Pull request merged**